### PR TITLE
Pipeline parallelism (linear)

### PR DESCRIPTION
**Pull request by gobbleturk**
Full PR with circular support is https://github.com/google/maxtext/pull/677 - we have decided to split up the PR into two

1. Without circular at first (This PR)
2. Adding circular

See https://github.com/google/maxtext/pull/677 for details on tests run in addition to the ones added in the PR. Note that without circular pipelining the bubble is usually large (or microbatch_size is small) so hard to get good performance.

### CODE REVIEW

File: MaxText/layers/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+  def get_iteration_inputs(self, loop_iteration, state_io, shift):
+    '''
+    Construct stages_in: the global array that is operated on for this iteration, shape same as shift=[stages, micro_size, sequence, embed]
+    This is almost a rotated version of the last outputs, except for the first stage which must grab a new batch from state_io
+    '''
+
+    # Setup potential input from state_io, which has a rotating microbatch index (size of microbatches_per_stage)
+    state_io_batch_idx = loop_iteration % self.microbatches_per_stage
+    first_stage_in = state_io[:,state_io_batch_idx]
```
Comment (jonb377) [inline MaxText/layers/pipeline.py:92]: @gobbleturk How do we prevent stage 0 from processing the output data during the bubble?

**Review by jonb377 — COMMENTED** [review]
Awesome work Matt! I'm most of the way through, will finish this up on Monday.

File: MaxText/layers/pipeline.py (hunk 1 above)
Comment (jonb377) [reply MaxText/layers/pipeline.py:92]: It occurred to me that any output data processed will be discarded while still in the `shift` buffer, so it's a non-issue. Let me know if that's a correct understanding.

File: MaxText/configs/base.yml [hunk 2]
```diff
@@ -110,9 +118,10 @@ jax_cache_dir: "~/jax_cache"
 hardware: 'tpu' # Supported hardware types are 'tpu', 'gpu', 'gpu_multiprocess' and 'cpu'
 
 # Parallelism
-mesh_axes: ['data', 'fsdp', 'fsdp_transpose', 'sequence', 'tensor', 'autoregressive']
+mesh_axes: ['stage', 'data', 'fsdp', 'fsdp_transpose', 'sequence', 'tensor', 'autoregressive']
 logical_axis_rules: [
                       ['activation_batch', ['data', 'fsdp', 'fsdp_transpose',]],
+                      ['activation_embed_and_logits_batch', ['stage', 'data', 'fsdp', 'fsdp_transpose']],
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/configs/base.yml:127]: Added comment both explaining this sharding and the importance of keeping stage first

File: MaxText/layers/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    new_state = _update_state_io(old_state_io, stream_slice, output)
+    
+    new_loop_state = {
+      "state_io": new_state,
+      "shift": new_shift,
+      "loop_iteration": loop_iteration + 1
+    }
+    return new_loop_state
+   
+  def permute_output_ms_dim(self, output):
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:184]: Spelled it out in the method name and in one other place where it was abbreviated to ms in comments

File: MaxText/layers/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+   '''Run one loop iteration - gets weights and inputs for each stage, run the stages in parallel, and update the loop state.'''
+   state_io = loop_state['state_io']
+   shift = loop_state["shift"]
+   loop_iteration = loop_state["loop_iteration"]
+
+   microbatch_ids = self.get_microbatch_ids(loop_iteration)
+
+   stages_inputs = self.get_iteration_inputs(loop_iteration, state_io, shift)
+   # We checkpoint stages_inputs since we are grabbing only one slice of the state_io, don't need to save the entire buffer.
+   stages_inputs = jax.ad_checkpoint.checkpoint_name(stages_inputs, 'iteration_input')
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:222]: Ran some experiments and found that this line actually does decrease the memory usage, I believe for the reason the comment states - checkpointing only the necessary inputs, although would need to dive deeper into the trace to confirm

experimental results
https://paste.googleplex.com/5940443468791808

File: MaxText/layers/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+      example_position = None
+    if segment_ids is not None:
+      segment_ids = segment_ids.reshape((self.config.num_pipeline_microbatches, self.microbatch_size, self.config.max_target_length))
+      example_segmentation = jax.lax.broadcast(segment_ids[0], [self.num_stages])
+    else:
+      example_segmentation = None
+
+    loop_state = self.init_states(inputs)
+    
+    total_iterations = self.config.num_pipeline_microbatches + self.num_stages  - 1
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:257]: Done!

File: MaxText/layers/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    else:
+      remat_policy = jax.checkpoint_policies.save_only_these_names('iteration_input')
+    run_one_iteration_rematted = nn.remat(
+      run_iteration_scannable,
+      prevent_cse=not self.config.scan_pipeline_iterations, # prevent_cse not used with scan
+      policy=remat_policy
+    )
+
+    # The scan cannot be used on init since it broadcasts the weights, which aren't yet initialized.
+    if self.config.scan_pipeline_iterations and not self.is_initializing():
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:290]: Removed.

File: MaxText/layers/pipeline.py (hunk 1 above)
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:92]: Yes that is the correct understanding! Added a comment here similar to your point

File: MaxText/configs/base.yml [hunk 7]
```diff
@@ -86,6 +86,14 @@ logits_via_embedding: False
 normalize_embedding_logits: True  # whether to normlize pre-softmax logits if logits_via_embedding is true
 logits_dot_in_fp32: True  # whether to use fp32 in logits_dense or shared_embedding dot product for stability
 
+# pipeline parallelism
+# The number of decoder layers is equal to the product of num_stages and num_layers_per_pipeline_stage (does not yet support circular pipelines).
+num_layers_per_pipeline_stage: 1
```
Comment (jonb377) [inline MaxText/configs/base.yml:91]: Do we need this to be specified or can it be derived as `num_layers / (ici_pipeline_parallelism * dcn_pipeline_parallelism)`?

**Review by jonb377 — APPROVED** [review]
LGTM! Thanks Matt 🔥

File: MaxText/layers/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    '''
+
+    # Shift is used to rotate the output of each pipeline into the input of the next
+    # shift has shape [num_stages, micro_size, sequence, embed]
+    shift = jnp.zeros((self.num_stages,) + inputs.shape[1:], dtype=inputs.dtype)
+    shift = nn.with_logical_constraint(shift, ("activation_stage", "activation_batch", "activation_length", "activation_embed"),rules=self.config.logical_axis_rules,mesh=self.mesh)
+
+    # state_io (state input output) at first holds all of the input batches, but also will hold the outputs as the pipeline runs/finishes
+    # state_io has shape [num_stages, microbatches/stages, micro_size, sequence, embed]
+    state_io = jnp.reshape(inputs, (self.num_stages, self.microbatches_per_stage) + inputs.shape[1:])
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers/pipeline.py:73]: It might be possible to have only one microbatch dimension for state_io, but the separation allows for easy manipulation of the desired sharding pattern (shard by stages) with minimal resharding required - The pattern is that each iteration we need to push a microbatch out of state_io and also one into state_io from stages - here we are guaranteed that the microbatch at the top of state_io is on the correct device(s) (stage 0 device(s)), and the microbatch to be pushed to the end is also at the correct device (corresponding to the last stage), and we only need to collective permute microbatches_per_stage at each iteration instead of all of them. This may be possible with just a single axis but would require complex bookkeeping/indexing

File: MaxText/configs/base.yml (hunk 7 above)
Comment (gobbleturk) [reply MaxText/configs/base.yml:91]: For the linear pipeline indeed this can be inferred as you show, however for circular there will be ambiguity (the product of num_layers_per_pipeline_stage and circ_repeats should be this value `num_layers / (ici_pipeline_parallelism * dcn_pipeline_parallelism)`
